namer/consul/src/main/scala/io/buoyant/namer/consul/SvcAddr.scala

```diff
@@ -76,29 +80,23 @@ private[consul] object SvcAddr {
             )
             stopped = true
             Future.Unit
-          case Throw(Failure(Some(err: ConnectionFailedException))) =>
-            log.warning(
-              "consul datacenter '%s' service '%s' retrying on connection" +
-                " failure: %s",
-              datacenter, key.name, err
-            )
-            // Drop the index, in case it's been reset by a consul restart
-            loop(None)
           case Throw(e) =>
             // an error occurred. if we've previously seen a good state, fall
             // back to that state; otherwise, fail.
             stats.errors.incr()
             lastGood match {
               case Some(addr) =>
-                // if we've already seen a good state, fall back to that and
-                // try again.
-                state() = addr
-                log.warning(
+                // do not update state if we were in good state before error
+                log.log(
+                  nonFatalLogLevel,
                   "consul datacenter '%s' service '%s' observation error %s," +
                     " falling back to last good state",
                   datacenter, key.name, e
                 )
-                loop(index0)
+                sleepAndDo(backoffs.getOrElse(apiBackoffs)) { nextBackoffs =>
+                  // retry and log subsequent errors as debug until state changes
+                  loop(None, Some(nextBackoffs), Level.DEBUG)
+                }
               case None =>
```

Comment:
I'm not sure that is makes sense to die here if there is no previous good state, especially given that we are removing the retry filter.  In the case of an error when there is no previous good state, I think we should not set `state()` (e.g. leave it as pending) and do a retry with backoff, just like the case above.  

@hawkw what do you think?

Reply:
I was having a specific case in mind: we have security scanner running periodically (I recall seeing someone else among linkerd users with similar setup). Every now and then scanner queries for some random addresses. I'd prefere those to not be accumulated over time causing unnecessary load.
Also requests to nonexistent DCs, I believe, shouldn't be retried (although, they were previously --- in infinite retry filter). It's seems unlikely, that there are envs where DC in consul may suddenly appear after request came and thus transitioning to pending with retries will likely give no benefit ever.

Reply:
Fair points.  It's hard to differentiate between an invalid request (such as from a security scanner or for a nonexistent DC) from a transient failure that should be retried.  Do you know what Linkerd's behavior is when it falls into this case?  Will it cache the Addr.Failed and continue to fail subsequent requests without retrying the consul API call?  If so, I don't think that's the behavior that we want since it would effectively make transient failures put Linkerd into a "stuck" state for that name.

Regarding unnecessary load, Linkerd will stop watching these names once they have been idle for 10 minutes, so the load won't be unbounded.  And hopefully the exponential backoff will mean that the added load from these API calls will be not too much.

Reply:
@adleong ugh, I didn't think of that. No I do not know what is the linkerd behavior.
I'll check this later.

Reply:
I looked, how Addr.Failed will be handled by Linkerd. Here what happens to `Addr.Failed` in case if mesh interpreter is used:

1. `SvcAddr`: returns `Addr.Failed`
2. `LookupCache`: transforms it into `Activity.Failed[NameTree[Name]]`. Value gets cached through Memoize (sic!)
3. `ConsulNamer` and farther: same
4. `mesh.InterpreterService.toBoundTreeRspEv`: transform to `VarEventStream.End(Throw(e))` (terminal event, containing either value or error)

  wire

5. `mesh.Client.streamActivity`: transforms End event into `Activity.Failed[NameTree[Name]]`
6. `Dst`: transforms to `Activity.Failed[BoundTree]`
7. `DynBoundFactory`: internal state transitions from Pending -> Failed. On `apply()` factory will return `Future.exception` (DunBoundFactory:81)
8. `ServiceFactoryCache`: wraps `DynBoundFactory` decorated with `NoBrokersAvailableException` handling into `IdlingFactory`
9. `IdlingFactory`: returns same `Future.exception`. Value remains in cache for TTL if idle
10. `RoutingService`: invokes `ServiceFactoryCache` and returns `Future.exception` as response. Every invocation resets idle timer, so as long as there are requests, `ServiceFactory` remains "stuck" in cache.

So it seems that
1. Even implementation of consul namer in master can put linkerd in "stuck" state (no previous state in namer, no index in response from consul). Obviously implementation from this PR will do this on any error. To get out of the "stuck" state linkerd has to be restarted or failing activity should be evicted from cache.
2. Considering that activities are memoized in namer, we also put namerd in "stuck" state. Looks like we should never interrupt a loop in activity (except for cases of explicit interruption). Theoretically[*], it can recover on its own, if all observes to activity are gone and activity is closed. Then the next observer will start the loop again.

Please, let me know if I missed something.

[*] during my experiments, I noticed that activity is not always closed as soon as all supposed observers are gone. I'd guess, there's another cache somewhere (maybe in http interface?). I had no time to dig into this however.

I'm going to change code to retry with back-off in case of exception.
I believe I should also change the case, when consul returns no index, but let me know if I misunderstood something. (Btw, why such case even exists? I understand, we can't do long poll on consul w/o index, but can't we just continue polling with some constant delay?)